libsql-wal/src/registry.rs

```diff
@@ -0,0 +1,205 @@
+        self.fs.create_dir_all(&path)?;
+        let dir = walkdir::WalkDir::new(&path).sort_by_file_name().into_iter();
+
+        let segments = SegmentList::default();
+        for entry in dir {
+            let entry = entry.map_err(|e| e.into_io_error().unwrap())?;
+            if entry
+                .path()
+                .extension()
+                .map(|e| e.to_str().unwrap() != "log")
+                .unwrap_or(true)
+            {
+                continue;
+            }
+
+            let file = self.fs.open(true, true, true, entry.path())?;
+
+            if let Some(sealed) =
+                SealedLog::open(file.into(), entry.path().to_path_buf(), Default::default())?
+            {
+                segments.push_log(sealed);
+            }
+        }
+
+        let db_file = self.fs.open(false, true, true, db_path)?;
+
+        // If this is a fresh database, we want to patch the header value for reserved space at the
+        // end of the file to store the replication index
+        let mut header: Sqlite3DbHeader = Sqlite3DbHeader::new_zeroed();
+        db_file.read_exact_at(header.as_bytes_mut(), 0)?;
+
+        let (db_size, start_frame_no) = segments
+            .with_head(|log| {
+                let header = log.header();
+                (header.db_size.get(), header.next_frame_no())
+            })
+            .unwrap_or((
+                header.db_size.get(),
+                // this should be the last frame_no in the db
+                NonZeroU64::new(1).unwrap(),
+            ));
+
+        let current_path = path.join(format!("{namespace}:{start_frame_no:020}.log"));
+
+        let log_file = self.fs.open(true, true, true, &current_path)?;
+
+        let current = arc_swap::ArcSwap::new(Arc::new(Log::create(
+            log_file,
+            current_path,
+            start_frame_no,
+            db_size,
+        )?));
+
+        // assert_eq!(header.reserved_in_page, 8, "bad db");
+
+        let shared = Arc::new(SharedWal {
+            current,
+            segments,
+            wal_lock: Default::default(),
+            db_file,
+            registry: self.clone(),
+            namespace: namespace.clone(),
+        });
+
+        openned.with_upgraded(|openned| {
+            openned.insert(namespace.clone(), shared.clone());
+        });
+
+        Ok(shared)
+    }
+
+    #[tracing::instrument(skip_all)]
+    pub fn swap_current(
+        &self,
+        shared: &SharedWal<FS>,
+        tx: &WriteTransaction<FS::File>,
+    ) -> Result<()> {
+        let before = Instant::now();
+        assert!(tx.is_commited());
+        // at this point we must hold a lock to a commited transaction.
+        // First, we'll acquire the lock to the current transaction to make sure no one steals it from us:
+        let lock = shared.wal_lock.tx_id.lock();
+        // Make sure that we still own the transaction:
+        if lock.is_none() || lock.unwrap() != tx.id {
```

Comment:
shouldn't this return an error?

Reply:
no, swapping the log is a best effort thing, if we don't have the lock, the next one that has the lock will try to swap. This is not an error